litestar/response/streaming.py

```diff
@@ -70,39 +89,51 @@ def __init__(
             media_type=media_type,
             status_code=status_code,
         )
-        self.iterator: AsyncIterable[str | bytes] | AsyncGenerator[str | bytes, None] = (
-            iterator if isinstance(iterator, (AsyncIterable, AsyncIterator)) else AsyncIteratorWrapper(iterator)
+
+        self.iterator: AsyncGenerator[str | bytes, None]
+
+        if isinstance(iterator, (AsyncGeneratorWrapper, AsyncGenerator)):
+            self.iterator = iterator
+        elif isinstance(iterator, (Iterable, Iterator)):
+            self.iterator = AsyncGeneratorWrapper(iterator)
+        elif isinstance(iterator, (AsyncIterable, AsyncIterator)):
+            self.iterator = async_iterator_to_generator(iterator)
+
+        self._original_generator = (
+            iterator.content_async_iterator if isinstance(iterator, ServerSentEventIterator) else self.iterator
         )
 
-    async def _listen_for_disconnect(self, cancel_scope: CancelScope, receive: Receive) -> None:
+    async def _listen_for_disconnect(self, disconnect_event: Event, receive: Receive) -> None:
         """Listen for a cancellation message, and if received - call cancel on the cancel scope.
 
         Args:
-            cancel_scope: A task group cancel scope instance.
+            disconnect_event: An event to set when a disconnect message is received.
             receive: The ASGI receive function.
 
         Returns:
             None
         """
-        if not cancel_scope.cancel_called:
-            message = await receive()
-            if message["type"] == "http.disconnect":
-                # despite the IDE warning, this is not a coroutine because anyio 3+ changed this.
-                # therefore make sure not to await this.
-                cancel_scope.cancel()
-            else:
-                await self._listen_for_disconnect(cancel_scope=cancel_scope, receive=receive)
-
-    async def _stream(self, send: Send) -> None:
+        while (message := await receive()) and not message["type"].endswith(".disconnect"):
+            pass
```

Comment:
I prefer the non-linlined version; It's more readable. With this, the intent isn't quite clear, as it looks just like a busy loop at first glance. Having a `while True`/`break` makes it more clear what's happening